keps/sig-network/3685-endpointslice-reconciler-to-staging/README.md

```diff
@@ -0,0 +1,154 @@
+
+### Non-Goals
+
+- A generic solution for reconciling fan-out style resources.
+
+## Proposal
+
+1. Add the staging repository (`staging/src/k8s.io/endpointslice`) to
+   `kubernetes/kubernetes` by following the steps in the [staging directory
+   README][staging-readme].
+2. Move the reconciler submodule and it’s [dependencies](###dependencies) to the
+   staging directory and update imports in `pkg/controller/endpointslice`. This
+   should be a single atomic change.
+    - Some methods need to be made public to allow for reconciler code to be
+      used as a library (at least `NewReconciler` and `Reconcile`).
+3. Allow for the value of the `endpointslice.kubernetes.io/managed-by` label to
+   be configurable by adding `managedBy` as an argument to `NewReconciler`.
+4. Split computing the diff (EndpointSlices to create, update, and delete) and
+   applying the diff into separate methods. This allows for controllers to
+   easily add metadata to the managed EndpointSlices (e.g. adding an
+   annotation).
+
+[staging-readme]: https://github.com/kubernetes/kubernetes/tree/master/staging "External Repository Staging Area"
+
+### Dependencies
+
+The following dependencies would also have to be moved to the staging repository
+(assume `staging_root = staging/src/k8s.io/endpointslice`):
+- `pkg/controller/endpointslice/metrics       => $staging_root/metrics`
+- `pkg/controller/endpointslice/topologycache	=> $staging_root/topologycache`
+- `pkg/controller/util/endpoint               => $staging_root/util`
+- `pkg/controller/util/endpointslice          => $staging_root/util`
+
+The following dependencies don’t have to be moved as there’s only one trivial
+function used from each:
+- `pkg/api/v1/pod`
+    - `IsPodReady` function can be moved to `component-helpers` as it’s also
+      used by `kubectl`.
+- `pkg/apis/core/v1/helper`
+    - `IsServiceIPSet` function can be moved to staging or mirrored.
+- `pkg/apis/discovery/validation`
+    - `ValidateEndpointSliceName` is a re-export of `NameIsDNSSubdomain` from
+      `k8s.io/apimachinery`, which is already in staging.
+
+### Risks and Mitigations
+
+Users might expect compatibility to be maintained for the library’s public API.
+This is a risk for all staging repositories, and the current mitigation strategy
+has been to clearly document that there are no compatibility guarantees in the
+README:
+
+> There are NO compatibility guarantees for this repository. It is in direct
+> support of Kubernetes, so branches will track Kubernetes and be compatible
+> with that repo. As we more cleanly separate the layers, we will review the
+> compatibility guarantee.
+
+## Design Details
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+## Production Readiness Review Questionnaire
+
+N/A
+
+## Implementation History
+
+## Drawbacks
```

Comment:
I think it's worth noting that this will add a bit of a maintenance cost to have code spread across different locations in the codebase.

Reply:
Good point, added.